lua-api-crates/table-funcs/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+
+    Ok(())
+}
+
+#[derive(Default, Debug, FromDynamic, ToDynamic, Clone, PartialEq, Eq, Copy)]
+enum ConflictMode {
+    /// Retain the existing value
+    Keep,
+    /// Take the latest value
+    #[default]
+    Force,
+    /// Raise an error
+    Error,
+}
+impl_lua_conversion_dynamic!(ConflictMode);
+
+#[derive(Default, Debug, FromDynamic, ToDynamic, Clone, PartialEq, Eq, Copy)]
+enum DepthMode {
+    /// Only look at the top level of tables
+    #[default]
+    Top,
+    /// Recursively go through tables
+    Deep,
+}
+impl_lua_conversion_dynamic!(DepthMode);
+
+// merge tables
+// (in case of overlap of the tables, we default to taking the key-value pair from the last table)
+// Note that we don't use a HashMap since we want to keep the order of the tables, which
+// can be useful in some cases
+fn extend<'lua>(
+    lua: &'lua Lua,
+    (array_of_tables, behavior): (Vec<Table<'lua>>, Option<ConflictMode>),
+) -> mlua::Result<Table<'lua>> {
+    let mut tbl_vec: Vec<(LuaValue, LuaValue)> = vec![];
+    for table in array_of_tables {
+        for pair in table.pairs::<LuaValue, LuaValue>() {
+            let (key, value) = pair?;
+            tbl_vec.push((key, value));
+        }
+    }
+    let tbl_len = tbl_vec.len();
+    // note we might allocate a bit too much here, but in many use cases we will be correct
+    let tbl: Table<'lua> = lua.create_table_with_capacity(0, tbl_len)?;
+
+    let behavior = behavior.unwrap_or_default();
+    for (key, value) in tbl_vec {
+        if !tbl.contains_key(key.clone())? {
+            tbl.set(key, value)?;
+        } else if behavior == ConflictMode::Force {
+            tbl.set(key, value)?;
+        } else if behavior == ConflictMode::Error {
+            return Err(mlua::Error::runtime(format!(
+                "The key {} is in more than one of the tables.",
+                key.to_string()?
+            )));
+        }
+    }
+
+    Ok(tbl)
+}
+
+// merge tables entrywise recursively
+// (in case of overlap of the tables, we default to taking the key-value pair from the last table)
+// Note that we don't use a HashMap since we want to keep the order of the tables, which
+// can be useful in some cases
+fn deep_extend<'lua>(
+    lua: &'lua Lua,
+    (array_of_tables, behavior): (Vec<Table<'lua>>, Option<ConflictMode>),
+) -> mlua::Result<Table<'lua>> {
+    let mut tbl_vec: Vec<(LuaValue, LuaValue)> = vec![];
+    for table in array_of_tables {
+        for pair in table.pairs::<LuaValue, LuaValue>() {
+            let (key, value) = pair?;
+            tbl_vec.push((key, value));
+        }
+    }
+    let tbl_len = tbl_vec.len();
+    // note we might allocate a bit too much here, but in many use cases we will be correct
+    let tbl: Table<'lua> = lua.create_table_with_capacity(0, tbl_len)?;
+
+    let behavior = behavior.unwrap_or_default();
+    for (key, value) in tbl_vec {
+        if !tbl.contains_key(key.clone())? {
+            tbl.set(key, value)?;
```

Comment:
> There's some inconsistency in behavior here:
> 
> ```lua
> local tbl1 = { a = 1 }
> local tbl2 = { a = {1,2,3} }
> 
> local res1 = extend_deep({tbl1, tbl2}, 'Force')
> local res2 = extend_deep({tbl2, tbl1}, 'Keep')
> ```
> 
> Even though both results superficially look the same:
> 
> ```
> {a = {1,2,3}}
> ```
> 
> `res1.a` is a clone of `tbl2.a` whereas `res2.a` is a reference to `tbl2.a`.
> 
> I think both should be clones.
> 
> What I think is that we should promote `wezterm.table.clone` to `wezterm.clone` and have it apply to an arbitrary lua value; that would be generally useful anyway and would simplify the logic in here.
> 
> Then we can call that here to make a deep copy of `value` before we assign it to the table.

I have made a few changes. I think both should be clones as you want now.

Reply:
I have cleanup up the `extend_deep` code more now. There were some edge cases, where it didn't behave as I would expect before. 

Now I do use clone, and I think the code itself is pretty clear now (as in, it is clear what the code is doing), but it did get a bit longer again because of different edge case behavior.